haystack/components/preprocessors/recursive_splitter.py

```diff
@@ -0,0 +1,208 @@
+
+    def _check_params(self):
+        if self.split_overlap < 0:
+            raise ValueError("Overlap must be greater than zero.")
+        if self.split_overlap >= self.split_length:
+            raise ValueError("Overlap cannot be greater than or equal to the chunk size.")
+        if not all(isinstance(separator, str) for separator in self.separators):
+            raise ValueError("All separators must be strings.")
+
+    @staticmethod
+    def _get_custom_sentence_tokenizer():
+        try:
+            from haystack.components.preprocessors.sentence_tokenizer import SentenceSplitter
+        except (LookupError, ModuleNotFoundError):
+            raise Exception("You need to install NLTK to use this function. You can install it via `pip install nltk`")
+        return SentenceSplitter()
+
+    def _apply_overlap(self, chunks: List[str]) -> List[str]:
+        """
+        Applies an overlap between consecutive chunks if the chunk_overlap attribute is greater than zero.
+
+        :param chunks: List of text chunks.
+        :returns:
+            The list of chunks with overlap applied.
+        """
+        overlapped_chunks = []
+        for idx, chunk in enumerate(chunks):
+            if idx == 0:
+                overlapped_chunks.append(chunk)
+                continue
+            overlap_start = max(0, len(chunks[idx - 1]) - self.split_overlap)
+            current_chunk = chunks[idx - 1][overlap_start:] + chunk
+            overlapped_chunks.append(current_chunk)
+        return overlapped_chunks
+
+    def _chunk_text(self, text: str) -> List[str]:
+        """
+        Recursive chunking algorithm that divides text into smaller chunks based on a list of separator characters.
+
+        It starts with a list of separator characters (e.g., ["\n\n", "\n", " ", ""]) and attempts to divide the text
+        using the first separator. If the resulting chunks are still larger than the specified chunk size, it moves to
+        the next separator in the list. This process continues recursively, progressively applying each specific
+        separator until the chunks meet the desired size criteria.
+
+        :param text: The text to be split into chunks.
+        :returns:
+            A list of text chunks.
+        """
+        if len(text) <= self.split_length:
+            return [text]
+
+        # type ignore below because we already checked that separators is not None
+        # try each separator
+        for separator in self.separators:  # type: ignore
+            if separator in "sentence":  # using nltk sentence tokenizer
+                sentence_with_spans = self.nltk_tokenizer.split_sentences(text)
+                splits = [sentence["sentence"] for sentence in sentence_with_spans]
+            else:
+                # split using the current separator
+                splits = text.split(separator) if not self.is_separator_regex else re.split(separator, text)
```

Comment:
apologies if this was already discussed, but is their a scenario where `re.split` can't always achieve the same thing as `text.split`? Basically just wondering why we don't always treat the separator as a regex?

Reply:
Given that separator will always be a single character, it seems the only differences are in performance, `text.split()` is slightly faster - I think that using regex for just a single character split might be a bit of a overhead - maybe we can remove it?

Reply:
Hmm I don't think so, if you look at LangChains implementation [here](https://api.python.langchain.com/en/latest/_modules/langchain_text_splitters/character.html#RecursiveCharacterTextSplitter) they provide a lot of out of the box sets of separators for different situations (e.g. Markdown, Different coding languages, etc.) and some of those are regexes. For example, 
```
        elif language == Language.RST:
            return [
                # Split along section titles
                "\n=+\n",
                "\n-+\n",
                "\n\\*+\n",
                # Split along directive markers
                "\n\n.. *\n\n",
                # Split by the normal type of lines
                "\n\n",
                "\n",
                " ",
                "",
            ]
```
I think it would be good for us to at least match their functionality.

Reply:
So I'd prefer for us to remove the `text.split()` and natively support splitting by regex.

Reply:
> Given that separator will always be a single character

sorry, this assumption I did is wrong, it can be a single character but also string (i.e.: regex expression), so there are definitely differences between both